### DISCUSSION TITLE
Add callback context pointer to stream_parser_t

### DISCUSSION THREAD

**Discussion by EricSidorov** [MaJerle/lwjson#14] [Ideas]
Hi! I was looking for a lightweight json lib for embedded devices and stumbled upon your great project.
I loved it and I had some improvement idea for the stream parser

It would be useful to pass user context pointer to the stream parser callback
for example if we'd like to parse
{"a":1,"b":{"c":true,"d":false}}
to a struct s_t
```
typedef struct _b
{
 char c;
 char d;
} b_t;

typedef struct _s
{
 int a
 b_t b
} s_t;
```
We could pass the struct object to the stream callback function by adding a void pointer to the stream_parset_t struct and its initializer function:
```
lwjsonr_t
lwjson_stream_init(lwjson_stream_parser_t* jsp, lwjson_stream_parser_callback_fn evt_fn, void *cb_ctx)
```
and then pass our struct to the callback function and fill its entries of the struct while parsing :
```
s_t out_struct;
void parse_s_callback_func(lwjson_stream_parser_t* jsp, lwjson_stream_type_t type)
{
 (s_t) *s = (s_t *) jsp->cb_ctx
 //Do all the parsing and fill the values of s
 ...
}

lwjson_stream_init(&jsp, parse_s_callback_func, &out_struct)
```
If this feature is desirable I'd be happy to implement it

**Comment by MaJerle** [comment]
Should such context be part of init function or rather of parser function?

**Reply by EricSidorov** [reply]
I think both ` lwjson_stream_parser_t.evt_fn` and it's context` lwjson_stream_parser_t.cb_ctx` should be initialized together in the init function
The function will be called with the context pointer in the parser function

**Comment by EricSidorov** [comment]
I implemented this feature in my fork and adjusted the stream example accordingly
https://github.com/EricSidorov/lwjson/tree/add_callback_context

**Comment by MaJerle** [comment]
My proposal only. Make a function that allows user to set user data which can be retrieved in the callback function.
In this way we do not break API compatibility

You thoughts?

**Reply by EricSidorov** [reply]
Good point
Updated my branch

**Reply by MaJerle** [reply]
Please go ahead and submit pull request to develop branch.
Please don't forget to run clang-format with your changes.

We can then comment add/update code directly on a request.
